imap_processing/codice/codice_l1a.py

```diff
@@ -741,6 +880,115 @@ def log_dataset_info(datasets: dict[int, xr.Dataset]) -> None:
+    # Chunk the data into each epoch/priority combination
+    for chunk in range(0, len(decompressed_data), num_priorities):
+        epoch_data = decompressed_data[chunk : chunk + num_priorities]
+
+        # The order of the priorities and data quality flags are unique to each
+        # epoch and can be gathered from the packet data
+        priority_order = packets.priority[chunk : chunk + num_priorities].data
+        data_quality = packets.suspect[chunk : chunk + num_priorities].data
+
+        # dict to hold the final data per epoch
+        data_per_epoch: dict[str, int | list[int | np.ndarray]] = {}
+
+        # For each epoch/priority combo, iterate over each event
+        for i, priority_num in enumerate(priority_order):
+            priority_data = epoch_data[i]
+
+            # Number of events and data quality can be determined at this stage
+            num_events = len(priority_data) // num_priorities
+            data_per_epoch[f"P{priority_num}_NumEvents"] = num_events
+            data_per_epoch[f"P{priority_num}_DataQuality"] = data_quality[i]
+
+            # Create blank array to store data for each epoch/priority/field
+            # combination
+            for field in constants.LO_DE_BIT_STRUCTURE:
+                data_per_epoch[f"P{priority_num}_{field}"] = []
+
+            # Iterate over each event
+            for event in [
+                priority_data[i * num_priorities : (i + 1) * num_priorities]
+                for i in range(num_events)
+            ]:
+                # Separate out each individual field from the bit string
+                # The fields are packed into the bit string in reverse order, so
+                # we need to back them out in reverse order
+                bit_string = "".join(f"{byte:08b}" for byte in event)
+                index = 0
+                for field_name, bit_length in reversed(
+                    constants.LO_DE_BIT_STRUCTURE.items()
+                ):
+                    data_per_epoch[f"P{priority_num}_{field_name}"].append(
+                        int(bit_string[index : index + bit_length], 2)
+                    )
+                    index += bit_length
+
+        # Append the epoch data to the final, restructured list of data
+        # For any data with less than 10000 events, pad the data array up to
+        # 10000
+        # TODO: Is this padding consistent with other instruments?
+        for i in range(num_priorities):
+            for field in constants.LO_DE_BIT_STRUCTURE:
+                if field not in ["Priority", "Spare"]:
+                    event_data = np.array(
+                        data_per_epoch[f"P{i}_{field}"], dtype=np.uint16
+                    )
+                    padding_size = 10000 - event_data.size
+                    padded_event_data = np.pad(
+                        event_data,
+                        (0, padding_size),
+                        mode="constant",
+                        constant_values=0,
```

Comment:
Is 0 the FILL_VAL here? Maybe we need to set it to whatever your FILL_VAL is in the attributes to notify folks this is an entry to be ignored.

Reply:
Good catch -- I will use the FILL_VAL here (which is `255`) so that it is more apparent.